src/pq/connection.cr

```diff
@@ -188,45 +189,44 @@ module PQ
       end
     end
 
+    def start_replication_frame_loop(publication_name : String, slot_name : String, &block : PG::Replication::Frame ->)
+      command = "START_REPLICATION SLOT #{slot_name} LOGICAL 0/0 (proto_version '1', binary 'true', publication_names '#{publication_name}')"
+      send_query_message command
+      loop do
+        break if soc.closed?
+        block.call PG::Replication::Frame.from_io(soc)
+      rescue e : IO::Error
+        raise e unless soc.closed?
+      end
+    end
+
     private def read_one_frame(frame_type)
       size = read_i32
       slice = read_bytes(size - 4)
       Frame.new(frame_type.not_nil!, slice) # .tap { |f| p f }
     end
 
     private def handle_async_frames(frame)
-      if frame.is_a?(Frame::ErrorResponse)
-        handle_error frame
-        true
-      elsif frame.is_a?(Frame::NotificationResponse)
-        handle_notification frame
-        true
-      elsif frame.is_a?(Frame::NoticeResponse)
-        handle_notice frame
-        true
-      elsif frame.is_a?(Frame::ParameterStatus)
-        handle_parameter frame
-        true
-      else
-        false
-      end
+      false
     end
 
-    private def handle_error(error_frame : Frame::ErrorResponse)
+    private def handle_async_frames(error_frame : Frame::ErrorResponse)
       expect_frame Frame::ReadyForQuery if @established
       notice_handler.call(error_frame.as_notice)
       raise PQError.new(error_frame.fields)
     end
 
-    private def handle_notice(frame : Frame::NoticeResponse)
-      notice_handler.call(frame.as_notice)
+    private def handle_async_frames(frame : Frame::NotificationResponse)
+      notification_handler.call(frame.as_notification)
+      true
     end
 
-    private def handle_notification(frame : Frame::NotificationResponse)
-      notification_handler.call(frame.as_notification)
+    private def handle_async_frames(frame : Frame::NoticeResponse)
+      notice_handler.call(frame.as_notice)
+      true
     end
 
-    private def handle_parameter(frame : Frame::ParameterStatus)
+    private def handle_async_frames(frame : Frame::ParameterStatus)
```

Comment:
The reason I updated this was because I was originally going to try to reuse `read_async_frame_loop`. That didn't work, but along the way I realized that the `if frame.is_a?` was basically a `case` statement, and even that was equivalent to just using an overloaded method. When I backed out my attempt, I just left it because I think this refactor still simplifies the code a bit.